src/SAWScript/MGU.hs

```diff
@@ -211,32 +210,36 @@ mgus ts1 ts2 =
   failMGU' $ "Wrong number of arguments. Expected " ++ show (length ts1) ++ " but got " ++ show (length ts2)
 
 -- Does not handle the case where t _is_ TyUnifyVar i; the caller handles that
-bindVar :: TypeIndex -> Type -> Either FailMGU Subst
-bindVar i t
-  | i `S.member` unifyVars t = failMGU' "occurs check failMGUs" -- FIXME: error message
-  | otherwise                = return (singletonSubst i t)
+bindVar :: Pos -> TypeIndex -> Type -> Either FailMGU Subst
+bindVar pos i t =
+  case M.lookup i $ unifyVars t of
+     Just otherpos ->
+       -- FIXME/XXX: this error message is better than the one that was here before
+       -- but still lacks
+       failMGU' $ "Occurs check failure: the type at " ++ show otherpos ++
+                  " appears within the type at " ++ show pos
+     Nothing ->
+       return (singletonSubst i t)
 
 -- }}}
 
 -- UnifyVars {{{
 
 class UnifyVars t where
-  unifyVars :: t -> S.Set TypeIndex
+  unifyVars :: t -> M.Map TypeIndex Pos
```

Comment:
It would be helpful to include a comment saying what the `Map` in the return type represents. (This isn't really your fault, since this function lacked a comment to begin with, but we might as well do better now that we need to update this code.)

Reply:
Agreed